pkg/ingester/retention.go

```diff
@@ -39,183 +65,322 @@ func defaultRetentionPolicy() retentionPolicy {
+func (*realFS) ReadDir(name string) ([]fs.DirEntry, error) { return os.ReadDir(name) }
+func (*realFS) RemoveAll(path string) error                { return os.RemoveAll(path) }
+
+type tenantBlock struct {
+	ID       ulid.ULID
+	TenantID string
+	Path     string
+	Uploaded bool
+}
+
+type fsBlockManager interface {
+	GetTenantIDs(ctx context.Context) ([]string, error)
+	GetBlocksForTenant(ctx context.Context, tenantID string) ([]*tenantBlock, error)
+	DeleteBlock(ctx context.Context, block *tenantBlock) error
+}
+
+type realFSBlockManager struct {
+	Root    string
+	Evictor blockEvicter
+	FS      fileSystem
+}
+
+func (bm *realFSBlockManager) GetTenantIDs(ctx context.Context) ([]string, error) {
+	tenantDirs, err := fs.ReadDir(bm.FS, bm.Root)
+	if err != nil {
+		return nil, err
+	}
+
+	tenantIDs := make([]string, 0)
+	for _, tenantDir := range tenantDirs {
+		if !tenantDir.IsDir() {
+			continue
+		}
+		tenantIDs = append(tenantIDs, tenantDir.Name())
+	}
+	return tenantIDs, nil
+}
+
+func (bm *realFSBlockManager) GetBlocksForTenant(ctx context.Context, tenantID string) ([]*tenantBlock, error) {
+	localDirPath := filepath.Join(bm.Root, tenantID, phlareDBLocalPath)
+	blockDirs, err := fs.ReadDir(bm.FS, localDirPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(bryan) manually read shipper file because of abstracted FS.
+	shipperPath := filepath.Join(localDirPath, shipper.MetaFilename)
+	bytes, err := fs.ReadFile(bm.FS, shipperPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var meta shipper.Meta
+	err = json.Unmarshal(bytes, &meta)
+	if err != nil {
+		return nil, err
+	}
+
+	uploadedBlockIDs := make(map[ulid.ULID]struct{}, len(meta.Uploaded))
+	for _, id := range meta.Uploaded {
+		uploadedBlockIDs[id] = struct{}{}
+	}
+
+	// Read blocks.
+	blocks := make([]*tenantBlock, 0)
+	for _, blockDir := range blockDirs {
+		if !blockDir.IsDir() {
+			continue
+		}
+
+		path := filepath.Join(localDirPath, blockDir.Name())
+		blockID, ok := block.IsBlockDir(path)
+		if !ok {
+			// A malformed/invalid ULID likely means that the directory is not a
+			// valid block, ignoring.
+			continue
+		}
+
+		_, uploaded := uploadedBlockIDs[blockID]
+		blocks = append(blocks, &tenantBlock{
+			ID:       blockID,
+			TenantID: tenantID,
+			Path:     path,
+			Uploaded: uploaded,
+		})
+	}
+	return blocks, nil
+}
+
+func (bm *realFSBlockManager) DeleteBlock(ctx context.Context, block *tenantBlock) error {
+	// TODO(bryan) Don't evict block if it's not marked as uploaded?
+
+	return bm.Evictor.evictBlock(block.TenantID, block.ID, func() error {
+		err := bm.FS.RemoveAll(block.Path)
+		switch {
 		case os.IsNotExist(err):
-			level.Warn(e.logger).Log("msg", "block not found on disk", "path", b.path)
-		default:
-			return fmt.Errorf("failed to delete block %q: %w", b.path, err)
+			return err
+		case err != nil:
+			return fmt.Errorf("failed to delete block: %q: %w", block.Path, err)
```

Comment:
I'm not sure I understand why we need to change this logic

Reply:
I changed it because the block manager no longer has a logger and it seemed relevant to log the fact we were skipping an error. In https://github.com/grafana/pyroscope/pull/2816/commits/312d6c0dfc45f2f6167d721ff34e717e5255a05f I updated the callsite of `DeleteBlock` to log and ignore the error, thus keeping the original logic intact.